logic/concludable/ConjunctionConcludable.java

```diff
@@ -112,21 +119,86 @@ public Value asValue() {
-            // Then build a Unification for each valid combination
-            return Stream.empty(); // TODO
+        public Stream<Map<Reference, Set<Reference>>> unify(HeadConcludable.Relation unifyWith) {
+            if (this.constraint().players().size() > unifyWith.constraint().players().size()) return Stream.empty();
+            Map<Reference, Set<Reference>> startingUnifier = new HashMap<>();
+            Optional<Map<Reference, Set<Reference>>> newUnifier;
+            if (constraint.owner().reference().isName()) {
+                newUnifier = updatedUnifier(this.constraint().owner(), unifyWith.constraint().owner(), startingUnifier);
+                if (newUnifier.isPresent()) startingUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            if (constraint.owner().isa().isPresent() && constraint.owner().isa().get().type().reference().isName()) {
+                assert unifyWith.constraint().owner().isa().isPresent();
+                newUnifier = updatedUnifier(this.constraint().owner().isa().get().type(),
+                        unifyWith.constraint().owner().isa().get().type(), startingUnifier);
+                if (newUnifier.isPresent()) startingUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            return matchRolesAndUnify(new ArrayList<>(this.constraint().players()), unifyWith.constraint().players(),
+                    startingUnifier, new HashSet<>()).stream();
+        }
+
+        private Set<Map<Reference, Set<Reference>>> matchRolesAndUnify(
+                List<RolePlayer> conjunctionRolePlayers, List<RolePlayer> headRolePlayers,
+                Map<Reference, Set<Reference>> unifier, Set<RolePlayer> visitedHeadRolePlayers) {
+            Set<Map<Reference, Set<Reference>>> allUnifiers = new HashSet<>();
+            if (conjunctionRolePlayers.isEmpty()) {
+                return set(unifier);
+            }
+            RolePlayer conj = conjunctionRolePlayers.remove(0);
+            for (RolePlayer head : headRolePlayers) {
+                if (visitedHeadRolePlayers.contains(head)) continue;
+                visitedHeadRolePlayers.add(head);
+                updatedUnifier(conj, head, unifier).ifPresent(newUnifier ->
+                        allUnifiers.addAll(matchRolesAndUnify(conjunctionRolePlayers,
+                                headRolePlayers, newUnifier, visitedHeadRolePlayers)));
+                visitedHeadRolePlayers.remove(head);
+            }
+            conjunctionRolePlayers.add(conj);
+            return allUnifiers;
+        }
+
+        Optional<Map<Reference, Set<Reference>>> updatedUnifier(
+                RolePlayer conjPlayer, RolePlayer headPlayer, Map<Reference, Set<Reference>> originalUnifier) {
+            if (!headPlayer.roleTypeHints().isEmpty() && !conjPlayer.roleTypeHints().isEmpty() &&
```

Comment:
can we make this more pseudo-codey by calling this `!contradictionExists(...)` or something?

Reply:
also, how about unification of this:
```
$x > 10;
```
and
```
$y < 5;
```

this seems like a contradiction that should be derived somehow?

Reply:
I did originally have a method for this but it's only used here, and we seem to prefer doing things inline when we can? Would be good to get a clarification on the style we prefer here.

Reply:
As to your second comment, I though about this. It seems that in this case it is not the unfier's job to prune at this point. The HeadConcludable will never be `$y<5`, as that contains an unnamed variable. It will be something like `$y<$num`. So the unifier should return `$x -> $y`, assuming the type hints are not disjoint.